Add Israel rail integration

shaiu opened the pull request:
## Proposed change
This proposed change introduces a new integration with the Israel Rail service (https://www.rail.co.il/) into Home Assistant. The integration aims to provide real-time information about the next three departure times from a specified origin station to a specified destination station.


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33685

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.


To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

shaiu committed:
Add Israel Rail integration

shaiu committed:
israel_rail tests

shaiu committed:
israel_rail tests

shaiu committed:
Merge branch 'dev' into shaiu-add-israel-rail

joostlek commented on homeassistant/components/israel_rail/__init__.py:
```diff
@@ -0,0 +1,60 @@
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
```
```suggestion
    hass: HomeAssistant, entry: ConfigEntry
```
I think we can use entry.runtime data here

shaiu replied:
cool
Done

joostlek commented on homeassistant/components/israel_rail/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+                    user_input[CONF_START],
+                    user_input[CONF_DESTINATION],
+                )
+            except Exception:
```
Can we catch more specific errors as well? So maybe when the service is down or something we want to display something else than if the input is incorrect

shaiu replied:
I checked and there isn't specific errors in the api (as for now)
I will reach out to the author or even suggest a change and add specific errors

joostlek commented on homeassistant/components/israel_rail/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+                _LOGGER.exception("Unknown error")
+                errors["base"] = "unknown"
+            if not errors:
+                unique_id = f"{user_input[CONF_START]} {user_input[CONF_DESTINATION]}"
```
Can you give an example of the start and destination?

shaiu replied:
These are station names of the railway. It's unique
like:
בית יהושע
תל אביב - סבידור מרכז

joostlek replied:
Can we get a list of all stations someone can choose from?

shaiu replied:
Yes I think that can be an option

shaiu replied:
Done

joostlek commented on homeassistant/components/israel_rail/coordinator.py:
```diff
@@ -0,0 +1,114 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class DataConnection(TypedDict):
+    """A connection data class."""
```
Why not make it a real dataclass

shaiu replied:
Done

joostlek commented on homeassistant/components/israel_rail/coordinator.py:
```diff
@@ -0,0 +1,114 @@
+        self._train_schedule = train_schedule
+        self._start = start
+        self._destination = destination
+        self.unique_id = f"{start} {destination}"
```
Not sure why we do this here

shaiu replied:
Removed

joostlek commented on homeassistant/components/israel_rail/sensor.py:
```diff
@@ -0,0 +1,102 @@
+    async_add_entities(entities)
+
+
+class IsraelRailEntitySensor(
```
Can we use entity descriptions?

shaiu replied:
Done

shaiu committed:
1. use entry.runtime
2. DataConnection - data class
3. remove unique id from coordinator
4. use EntityDescription

shaiu committed:
Merge remote-tracking branch 'origin/shaiu-add-israel-rail' into shaiu-add-israel-rail

shaiu committed:
add a list of stations in user form

joostlek commented on tests/components/israel_rail/conftest.py:
```diff
@@ -0,0 +1,182 @@
+            "homeassistant.components.israel_rail.TrainSchedule",
+            return_value=mock_israelrail_api,
+        ),
+        patch(
+            "homeassistant.components.israel_rail.coordinator.TrainSchedule",
+            return_value=mock_israelrail_api,
+        ),
```
```suggestion
```

shaiu replied:
Removed

shaiu committed:
1. extend ConfigEntry
2. remove unused pop
3. use IsraelRailSensorEntityDescription to have only one kind of Sensor
4. add test for already configured
5. use snapshot in test

shaiu committed:
change user step description

shaiu committed:
Merge branch 'dev' into shaiu-add-israel-rail

shaiu commented on homeassistant/components/israel_rail/__init__.py:
```diff
@@ -0,0 +1,57 @@
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Done

joostlek replied:
Not yet, you still need to use the `IsraelRailConfigEntry` here

shaiu commented on homeassistant/components/israel_rail/__init__.py:
```diff
@@ -0,0 +1,57 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
```
removed

shaiu commented on homeassistant/components/israel_rail/sensor.py:
```diff
@@ -0,0 +1,164 @@
+    entities = [
+        IsraelRailDepartureSensor(coordinator, description, unique_id)
+        for description in DEPARTURE_SENSORS
+    ] + [
```
Changed

shaiu commented on homeassistant/components/israel_rail/sensor.py:
```diff
@@ -0,0 +1,164 @@
+
+class IsraelRailSensor(IsraelRailEntitySensor):
+    """Implementation of a Israel Rail other sensor."""
+
```
Done.
Consolidated to one sensor

shaiu commented on homeassistant/components/israel_rail/strings.json:
```diff
@@ -0,0 +1,45 @@
+          "from": "Start station",
+          "to": "End station"
+        },
+        "description": "Provide start and end station for your connection\n\nCheck https://www.rail.co.il for valid stations.",
```
Done

shaiu commented on tests/components/israel_rail/conftest.py:
```diff
@@ -0,0 +1,182 @@
+
+@pytest.fixture
+def mock_israelrail():
+    """Build a fixture for the 17Track API."""
```
Changed to israelrail

shaiu commented on tests/components/israel_rail/conftest.py:
```diff
@@ -0,0 +1,182 @@
+"""Configuration for 17Track tests."""
```
Changed to israelrail

shaiu commented on tests/components/israel_rail/conftest.py:
```diff
@@ -0,0 +1,182 @@
+    ),
+]
+
+trains_wrong_format = [
```
Done

shaiu commented on tests/components/israel_rail/test_config_flow.py:
```diff
@@ -0,0 +1,65 @@
+
+    assert result2["type"] is FlowResultType.CREATE_ENTRY
+    assert result2["title"] == "start destination"
+    assert result2["data"] == {
```
Done

shaiu commented on tests/components/israel_rail/test_config_flow.py:
```diff
@@ -0,0 +1,65 @@
+        failed_result["flow_id"],
+        VALID_CONFIG,
+    )
+    await hass.async_block_till_done()
```
Removed

shaiu commented on tests/components/israel_rail/test_config_flow.py:
```diff
@@ -0,0 +1,65 @@
+        data=VALID_CONFIG,
+    )
+
+    assert failed_result["errors"] == {"base": "unknown"}
```
Added assert

shaiu commented on tests/components/israel_rail/test_config_flow.py:
Added

shaiu commented on tests/components/israel_rail/test_sensor.py:
```diff
@@ -0,0 +1,91 @@
+    assert hass.states.get("sensor.mock_title_duration")
+    assert hass.states.get("sensor.mock_title_transfers")
+    assert hass.states.get("sensor.mock_title_platform")
+    assert hass.states.get("sensor.mock_title_train_number")
```
moved to snapshot

shaiu commented on tests/components/israel_rail/test_sensor.py:
```diff
@@ -0,0 +1,91 @@
+) -> None:
+    """Ensure nothing is created when config is wrong."""
+    await init_integration(hass, mock_config_entry)
+    assert not hass.states.async_entity_ids("sensor")
```
added excpetion

joostlek commented on homeassistant/components/israel_rail/strings.json:
```diff
@@ -0,0 +1,45 @@
+      "departure2": {
+        "name": "Departure +2"
+      },
+      "duration": {
+        "name": "Duration"
+      },
```
Can be removed

shaiu replied:
removed

joostlek commented on homeassistant/components/israel_rail/strings.json:
```diff
@@ -0,0 +1,45 @@
+        "name": "Platform"
+      },
+      "train_number": {
+        "name": "Train Number"
```
```suggestion
        "name": "Train number"
```

shaiu replied:
fixed

joostlek commented on tests/components/israel_rail/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    )
+
+    assert failed_result["errors"] == {"base": "unknown"}
+    assert failed_result["type"] == FlowResultType.FORM
```
```suggestion
    assert failed_result["type"] is FlowResultType.FORM
```

shaiu replied:
done

joostlek commented on tests/components/israel_rail/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    entry.add_to_hass(hass)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

shaiu replied:
done

joostlek commented on tests/components/israel_rail/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+
+async def test_flow_fails(
+    hass: HomeAssistant,
+    mock_israelrail: AsyncMock,
```
also add the mock_setup_entry here

shaiu replied:
added

joostlek commented on tests/components/israel_rail/test_sensor.py:
```diff
@@ -0,0 +1,98 @@
+) -> None:
+    """Ensure everything starts correctly."""
+    await init_integration(hass, mock_config_entry)
+    entity_entries = er.async_entries_for_config_entry(
+        entity_registry, mock_config_entry.entry_id
+    )
+    assert entity_entries
+    assert len(hass.states.async_entity_ids()) == 7
+    for entity_entry in entity_entries:
+        state = hass.states.get(entity_entry.entity_id)
+        assert state == snapshot(name=f"{entity_entry.entity_id}")
```
just use the `snapshot_platform` helper

shaiu replied:
cool, didn't know it exists

shaiu replied:
Done

shaiu committed:
1. ConfigEntry[IsraelRailDataUpdateCoordinator]
2. remove redundant attributes
3. use snapshot_platform helper

shaiu committed:
Merge branch 'dev' into shaiu-add-israel-rail

shaiu commented on homeassistant/components/israel_rail/__init__.py:
```diff
@@ -0,0 +1,66 @@
+    coordinator: IsraelRailDataUpdateCoordinator
+
+
+type IsraelRailConfigEntry = ConfigEntry[IsraelRailEntryData]
```
Done

shaiu commented on homeassistant/components/israel_rail/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/israel_rail",
+  "iot_class": "cloud_polling",
+  "loggers": ["israel_rail"],
```
changed

shaiu commented on homeassistant/components/israel_rail/sensor.py:
```diff
@@ -0,0 +1,137 @@
+    return lambda data: {
+        "start": data[i].start,
+        "destination": data[i].destination,
+    }
```
removed

joostlek commented on tests/components/israel_rail/test_sensor.py:
```diff
@@ -0,0 +1,92 @@
+    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
+
+
+async def test_invalid_config(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+    mock_israelrail: AsyncMock,
+) -> None:
+    """Ensure nothing is created when config is wrong."""
+    mock_israelrail.return_value.query.side_effect = Exception("error")
+    await init_integration(hass, mock_config_entry)
+    assert not hass.states.async_entity_ids("sensor")
```
Let's replace this test with a test in `test_init.py` that starts the same but then checks the state of the `mock_config_entry`

shaiu replied:
Done

shaiu committed:
remove attr

joostlek commented on tests/components/israel_rail/test_sensor.py:
```diff
@@ -0,0 +1,92 @@
+    await goto_future(hass, freezer)
+    assert len(hass.states.async_entity_ids()) == 7
+    departure_sensor = hass.states.get("sensor.mock_title_departure")
+    assert departure_sensor.state == "unavailable"
```
```suggestion
    assert departure_sensor.state == STATE_UNAVAILABLE
```

shaiu replied:
Done

shaiu committed:
remove attr

shaiu committed:
move test to test_init.py

shaiu committed:
Merge branch 'dev' into shaiu-add-israel-rail

joostlek committed:
Merge branch 'dev' into shaiu-add-israel-rail

joostlek committed:
Fix

joostlek committed:
Fix

joostlek committed:
Fix

joostlek committed:
Fix

joostlek reviewed (changes requested):
There's a timezone bug that is affecting the tests, please check it out

shaiu committed:
fix timezone

shaiu committed:
fix

shaiu committed:
Merge branch 'dev' into shaiu-add-israel-rail

shaiu committed:
fix

MartinHjelmare commented on homeassistant/components/israel_rail/coordinator.py:
```diff
@@ -0,0 +1,113 @@
+    transfers: int
+
+
+def calculate_duration_in_seconds(start_time: str, end_time: str) -> int | None:
```
We don't allow calculating state. Sensors should just return the native values of the API as appropriate according to the Home Assistant entity design guidelines.

shaiu replied:
Removed

MartinHjelmare commented on homeassistant/components/israel_rail/sensor.py:
```diff
@@ -0,0 +1,125 @@
+
+SENSORS: tuple[IsraelRailSensorEntityDescription, ...] = (
+    IsraelRailSensorEntityDescription(
+        key="duration",
```
Please remove this sensor unless the API can provide it directly.

shaiu replied:
Removed

MartinHjelmare commented on homeassistant/components/israel_rail/coordinator.py:
```diff
@@ -0,0 +1,113 @@
+    departure: datetime | None
+    duration: int | None
+    platform: str
+    remaining_time: str
```
Remaining time is never used.

shaiu replied:
Removed

MartinHjelmare commented on homeassistant/components/israel_rail/sensor.py:
```diff
@@ -0,0 +1,125 @@
+        value_fn=lambda data_connection: data_connection.platform,
+    ),
+    IsraelRailSensorEntityDescription(
+        key="transfers",
```
Transfers should probably just be trains without any deduction of 1.

shaiu replied:
But that would be wrong. If there is only one train from place A to B the number of transfers is 0

MartinHjelmare replied:
Right. I mean that we should rename this sensor to `trains` and just have it return the original count of trains.

shaiu replied:
Done

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
